telemetry-aware-scheduling/pkg/strategies/labeling/strategy.go

```diff
@@ -0,0 +1,140 @@
+// This label can then be used externally to act on the strategy violation.
+package labeling
+
+import (
+	"fmt"
+
+	"github.com/intel/platform-aware-scheduling/telemetry-aware-scheduling/pkg/cache"
+	"github.com/intel/platform-aware-scheduling/telemetry-aware-scheduling/pkg/strategies/core"
+	telempol "github.com/intel/platform-aware-scheduling/telemetry-aware-scheduling/pkg/telemetrypolicy/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+)
+
+// StrategyType is set to "labeling".
+const (
+	StrategyType = "labeling"
+)
+
+// Strategy type for labeling from a single policy.
+type Strategy telempol.TASPolicyStrategy
+
+// StrategyType returns the name of the strategy type. This is used to place it in the registry.
+func (d *Strategy) StrategyType() string {
+	return StrategyType
+}
+
+type ruleResult struct {
+	rule     telempol.TASPolicyRule
+	quantity resource.Quantity
+}
+
+type violationResultType struct {
+	ruleResults []ruleResult
+}
+
+// Violated checks if the strategy is violated by searching for nodes that have metrics that don't accord with
+// the target in labeling strategy.
+// Returns a map of nodeNames as key with a slice of violated rules and metric quantities in the result type.
+func (d *Strategy) Violated(cache cache.Reader) map[string]interface{} {
+	violatingNodes := map[string]interface{}{}
+
+	for _, rule := range d.Rules {
+		nodeMetrics, err := cache.ReadMetric(rule.Metricname)
+		if err != nil {
+			klog.V(2).InfoS(err.Error(), "component", "controller")
+
+			continue
+		}
+
+		for nodeName, nodeMetric := range nodeMetrics {
+			msg := fmt.Sprint(nodeName+" "+rule.Metricname, " = ", nodeMetric.Value.AsDec())
+			klog.V(4).InfoS(msg, "component", "controller")
+
+			if core.EvaluateRule(nodeMetric.Value, rule) {
+				msg := fmt.Sprintf(nodeName + " violating " + d.PolicyName + ": " + ruleToString(rule))
+				klog.V(2).InfoS(msg, "component", "controller")
+
+				if _, ok := violatingNodes[nodeName]; !ok {
+					violatingNodes[nodeName] = &violationResultType{}
+				}
+
+				res, ok := violatingNodes[nodeName].(*violationResultType)
+				if !ok {
+					klog.Error("unexpected type")
+
+					continue
+				}
+
+				res.ruleResults = append(res.ruleResults, ruleResult{rule: rule, quantity: nodeMetric.Value})
+				if len(res.ruleResults) > 0 {
+					for _, ruleRes := range res.ruleResults {
+						klog.V(2).Infof("Violated rules: %v", ruleToString(ruleRes.rule))
+					}
+				}
+			}
+		}
+	}
+
+	return violatingNodes
+}
+
+// ruleToString returns the rule passed to it as a single string.
+func ruleToString(rule telempol.TASPolicyRule) string {
+	return fmt.Sprintf("%v %v %v %v", rule.Metricname, rule.Operator, rule.Target, rule.Labels)
+}
+
+func equalRules(a, b *telempol.TASPolicyRule) bool {
+	if a.Metricname != b.Metricname {
+		return false
+	}
+
+	if a.Target != b.Target {
+		return false
+	}
+
+	if a.Operator != b.Operator {
+		return false
+	}
+
```

Comment:
Is it possible to use one IF condition with logical 'OR' operators here ? It may just save some of the lines on the code ?

    var bVal bool = True   // default is false, so setting it true
    if (a.Metricname != b.Metricname) || (a.Operator != b.Operator) || (a.Target != b.Target)
    {
          bVal  = False
    }

   for j := range a.Labels {
		if a.Labels[j] != b.Labels[j] {
			bval = False
		}
	}
    return bVal

Reply:
done in separate commit

Reply:
Looks good to me.